sips/sip-030/sip-030-wallet-interface.md

```diff
@@ -0,0 +1,571 @@
+
+Result properties
+
+- `signature`: `string` hex-encoded
+- `publicKey`: `string` hex-encoded
+
+#### Method `stx_signStructuredMessage`
+
+Parameter properties
+
+- `message`: `ClarityValue`
+  - where `ClarityValue` is `string | object` hex-encoded or JSON representation
+- `domain`: `ClarityTupleValue`
+  - where `ClarityTupleValue` (defined by SIP-018) is a Clarity "tuple" as `string | object` hex-encoded or JSON representation
+
+Result properties
+
+- `signature`: `string` hex-encoded
+- `publicKey`: `string` hex-encoded
+
+#### Method `stx_getAddresses`
+
+> **Note**: This method can be used similarly to legacy "connect" methods, where the first account selection also acts as an approval of the site/domain.
+
+Result properties
+
+- `addresses`: `{}[]`
+  - `address`: `string` address, Stacks c32-encoded
+  - `publicKey`: `string` hex-encoded
+
+#### Method `stx_getAccounts`
+
+> **Comment**: This method is similar to `stx_getAddresses`.
+> It was added to provide better backwards compatibility for applications using Gaia.
+
+> **Note**: This method can be used similarly to legacy "connect" methods, where the first account selection also acts as an approval of the site/domain.
+
+Result properties
+
+- `accounts`: `{}[]`
+  - `address`: `string` address, Stacks c32-encoded
+  - `publicKey`: `string` hex-encoded
+  - `gaiaHubUrl`: `string` URL
+  - `gaiaAppKey`: `string` hex-encoded
+
+#### Method `stx_getNetworks`
+
+Result properties
+
+- `active`: `string` network identifier
+- `networks`: `{}[]`
+  - `id`: `string`
+  - `chainId`: `number`
+  - `transactionVersion`: `number`
+
+#### Method `stx_updateProfile`
+
+Parameter properties
+
+- `profile`: `object` Schema.org Person object[^13]
+
+Result properties
+
+- `profile`: `object` updated Schema.org Person object[^13]
+
+### Listeners
+
+In addition to the request interface, event listeners may be provided via the `.listen` method.
+
+- `provider.listen(event: string, listener: (...args: any[]) => void): Function`
+
+> `provider.listen` should return a "unlisten" function, which can be called to remove the listener.
+
+The event name should be closer to nouns than verbs and doesn't use the `on` prefix from DOM naming conventions.
+
+#### Event `stx_accountChange`
+
+- `listener: (accounts: {}[]) => void`
+
+> `accounts` as defined above in `stx_getAccounts`.
+> The first account is considered the default account (and may be the only "active" account in a wallet).
+
+#### Event `stx_networkChange`
+
+- `listener: (data: GetNetworksResult) => void`
+  - where `GetNetworksResult` matches the `stx_getNetworks` result type
+
+> This event may be emitted when the active network changes and on first load of the wallet/site.
+
+### Error Codes
+
+Errors thrown by request methods should match existing JSON-RPC 2.0 error codes.
+This way, the user or an intermediary library can handle them in a standardized way.
+Otherwise, no additional error codes are defined in this SIP.
```

Comment:
Common errors like "user denied request" should have a defined error code
Leather uses 4001.

Reply:
Added